agent/consul/leader.go

```diff
@@ -77,7 +77,10 @@ func (s *Server) monitorLeadership() {
 				leaderLoop.Add(1)
 				go func(ch chan struct{}) {
 					defer leaderLoop.Done()
-					s.leaderLoop(ch)
+					err := s.leaderLoop(ch)
+					if err != nil {
+						s.leadershipTransfer()
```

Comment:
What if leadership transfer fails? We exit the leader loop anyway and no longer act as the leader, but does raft still think we are the leader? That seems like a bad case to be in - roughly the same as the bug this is meant to be fixing although we at least made an attempt at telling raft we didn't want to be leader any more.

I can think of a couple of possible options here:
 - Make a method in Raft lib that allows the leader to "StepDown" - it will stop running heartbeats etc. but not stop being a voter - basically, force a leader into follower state.
 - Keep trying the leadership Transfer indefinitely because the cluster is broken until it works anyway.
 - Keep retrying for a limited length of time and then crash the whole process to force a step down.

Reply:
I made a couple of changes to my PR, so that `leaderLoop` is only left if `leadershipTransfer` was successful. I think that mitigates the issues you are talking about.

What happens now is that in case `leadershipTransfer` fails, consul stays in `leaderLoop` and waits until `ReconcileInterval` to retry to `establishLeadership` (and to `transferLeadership` in case it fails).

Reply:
I think this is OK although `ReconcileInterval` is 60 seconds which seems a lot to wait around for when we know the cluster is down. 🤔

Should we instead consider looping indefinitely retrying every 5 seconds or something?

I think this is way better than before and eventually it should recover so it's good just wondering if it's easy to make that better quicker?

Reply:
I reset the `interval` to 5 seconds when transfer leadership fails so that it retries establishLeadership again faster than before.